File: DSCI 553/Homework 2/task2.scala
import org.apache.log4j.{Level, Logger}
import org.apache.spark.sql.SparkSession

import scala.collection.mutable.ListBuffer
import scala.collection.{immutable, mutable}
import java.io.PrintWriter

object task2 {
  def main(args: Array[String]) {
    Logger.getLogger("org").setLevel(Level.ERROR)

    val ss = SparkSession
      .builder()
      .appName("task1")
      .config("spark.master", "local[*]")
      .getOrCreate()

    val k = Integer.parseInt(args(0))
    val support = Integer.parseInt(args(1))
    val input_file = args(2)
    val output_path = args(3)

    val sc = ss.sparkContext

    val start_time = System.nanoTime()

    val textRdd = sc.textFile(input_file)
    val rdd = textRdd.filter(line => line != "user_id,business_id")
      .map(line => line.split(","))
      .map(x => ( x(0), immutable.Set(x(1)) ) )
      .reduceByKey((x, y) => x | y)
      .filter( x => x._2.size > k )
      .map(x => x._2)


    val count = rdd.count()

    val frequent_candidates = rdd.mapPartitions( x => getItemSet(x, count, support) )
      .distinct().collect().toList

    val frequent_itemsets = rdd.mapPartitions( x => countFrequentItemset(x, frequent_candidates) )
      .reduceByKey(_+_)
      .filter( x => x._2 >= support )
      .map( x => x._1 ).collect().toList

    val ans = "Candidates:\n"+getOutputInFormat(frequent_candidates)+"\n\n"+"Frequent Itemsets:\n"+getOutputInFormat(frequent_itemsets)

    new PrintWriter(output_path) { write(ans); close }
    println("Duration:"+( 1.0*(System.nanoTime()-start_time)/1000000000 ))
  }

  def getOutputInFormat( itemsets : List[ Set[String] ] ): String = {
    val length_dict = mutable.HashMap[Int, ListBuffer[List[String]]]()
    var mx = 0
    var temp = mutable.ListBuffer[List[String]]()
    for (item <- itemsets) {
      temp = length_dict.getOrElse(item.size, mutable.ListBuffer[List[String]]())
      temp.append(item.toList.sorted)
      length_dict(item.size) = temp
      if (item.size > mx) {mx = item.size}
    }

    implicit val lengthContentOrdering: Ordering[List[String]] = new Ordering[List[String]] {
      override def compare(a: List[String], b: List[String]): Int = {
        for ((x, y) <- a zip b) {
          val compare = x.compareTo(y)
          if (compare != 0) {
            return compare
          }
        }
        0
      }
    }

    var ans = ListBuffer[String]()

    for ( i <- 1 to mx ){
      length_dict(i) = length_dict(i).sorted( lengthContentOrdering )
      var temp = ListBuffer[String]()
      for (list <- length_dict(i) ){
        temp.append( "(" + list.map(x => "\'"+x+"\'").mkString(", ") + ")" )
      }
      ans.append(temp.mkString(","))
    }

    ans.mkString("\n\n")
  }

  def countFrequentItemset(partition: Iterator[Set[String]],
                           frequent_candidates: List[ Set[String] ]): Iterator[ Tuple2[Set[String], Int] ] = {
    val count = mutable.HashMap[ Set[String], Int ]()
    val baskets = partition.toList
    for (basket <- baskets){
      for ( itemset <- frequent_candidates ) {
        if ( itemset.subsetOf(basket) ) {
          count(itemset) = count.getOrElse(itemset, 0) + 1
        }
      }
    }

    val freq_count = mutable.Set[ Tuple2[ Set[String], Int ] ]()
    for ( itemset <- frequent_candidates ){
      freq_count.add( (itemset, count.getOrElse(itemset, 0)) )
    }
    freq_count.iterator
  }

  def getItemSet(partition: Iterator[Set[String]],
                 cnt: Long, support: Integer ): Iterator[ Set[String] ] = {
    val baskets = partition.toList
    var singletons = mutable.Set[String]()
    for (basket <- baskets) {
      singletons = singletons.union(basket)
    }
    val threshold = (1.0 * support * baskets.length) / cnt

    val candidates = mutable.Set[ Set[String] ]()
    for (c <- singletons) {
      candidates.add(Set(c))
    }

    val frequent_candidates = mutable.Set[Set[String]]()
    var cand_size = 1
    while (candidates.size > 0) {
      var count = mutable.HashMap[Set[String], Int]()
      for (basket <- baskets) {
        for (candidate <- candidates) {
          if (candidate.subsetOf(basket)) {
            count(candidate) = count.getOrElse(candidate, 0) + 1
          }
        }
      }

      val frequent_itemset = mutable.Set[Set[String]]()
      for (candidate <- candidates) {
        if (count.getOrElse(candidate, 0) >= threshold) {
          frequent_itemset.add(candidate)
          frequent_candidates.add(candidate)
        }
      }

      candidates.clear()
      cand_size = cand_size + 1
      for (freq_items <- frequent_itemset.toList.combinations(2)) {
        val new_cand = freq_items(0).union(freq_items(1))
        if (new_cand.size == cand_size) {
          candidates.add(new_cand)
        }
      }
    }
    frequent_candidates.iterator
  }
}

File: DSCI 553/Homework 2/task2.py
from pyspark import SparkContext
from itertools import combinations
from collections import defaultdict
from operator import add
import sys
from time import time

#python task2.py 70 50 "datasets/ub.csv" output2.txt
sc = SparkContext("local[*]", "homework2_task2")
args = sys.argv

start = time()
k = int(args[1].strip())

rdd = sc.textFile(args[3].strip()) \
        .filter(lambda x: x.split(',')[0] != 'user_id') \
        .map(lambda x: tuple(x.split(','))) \
        .map(lambda x: (x[0], frozenset({x[1]})) ) \
        .reduceByKey(lambda x, y : x|y ) \
        .filter(lambda x: len(x[1]) > k) \
        .map( lambda x: x[1] )

count = rdd.count()
support = int(args[2].strip())

def getOutputInFormat( itemset ):
    length_dict, mx = defaultdict(list), 0
    for item in itemset:
        length_dict[ len(item) ].append(tuple(sorted(item)))
        mx = max(mx, len(item))
    
    output = []
    for length in range(1, mx+1):
        vals = length_dict[length]
        if length == 1:
            output.append(",".join( sorted([f"('{v[0]}')" for v in vals]) ) )
        else:
            output.append(",".join( sorted([f"{v}" for v in vals]) ) )
    return "\n\n".join(output)

def getItemSet(partition, cnt, support):
    """Uses Apriori without any improvement"""
    baskets = []
    
    singletons = set()
    for basket in partition: 
        baskets.append( basket )
        singletons.update( basket )
    
    threshold = (support*len(baskets))/cnt
    
    #convert all candidates to tuple
    candidates = { frozenset( (c,) ) for c in singletons }
    frequent_candidates = set()
    cand_size = 1
    while len(candidates) > 0:
        #calculating count of each candidate from all baskets
        count = defaultdict(int)
        for basket in baskets:
            for candidate in candidates:
                if candidate.issubset(basket):
                    count[candidate] += 1

        #filtering candidates based on threshold
        frequent_itemset = set()
        for candidate in candidates:
            if count[candidate] >= threshold:
                frequent_itemset.add( candidate )
                frequent_candidates.add( candidate )
        
        #combining only frequeunt candidates to form new candidates of size+1
        candidates.clear()
        cand_size += 1
        for freq1, freq2 in combinations(frequent_itemset, 2):
            new_cand = freq1.union(freq2)
            valid = True
            if len(new_cand) != cand_size: continue
            for x in combinations(new_cand, cand_size-1):
                if frozenset(x) not in frequent_candidates:
                    valid = False
                    break
            if valid: candidates.add( new_cand )
    
    return frequent_candidates

def countFrequentItemset(partition, frequent_candidates):
    count = defaultdict(int)
    baskets = []
    for basket in partition:
        baskets.append(basket)
        
    for basket in baskets:
        for itemset in frequent_candidates:
            if itemset.issubset(basket):
                count[ itemset ]+=1
    
    freq_count = []
    for itemset, cnt in count.items():
        freq_count.append( (itemset, cnt) )
    return freq_count

frequent_candidates = rdd.mapPartitions( lambda x: getItemSet(x, count, support) ) \
                        .distinct().collect()
                        
frequent_itemsets = rdd.mapPartitions(lambda x: countFrequentItemset(x, frequent_candidates)) \
        .reduceByKey(add) \
        .filter( lambda x: x[1] >= support ) \
        .map(lambda x: x[0]) \
        .collect()

with open( args[4].strip(), "w" ) as out:
    print("Candidates:", file=out)
    print( getOutputInFormat(frequent_candidates), file=out )
    print(file=out)
    print("Frequent Itemsets:", file=out)
    print( getOutputInFormat(frequent_itemsets), file=out)

print(f"Duration: {time()-start}", file=sys.stdout)